third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -268,128 +249,108 @@ bool canNarrowOffset(Value baseOffset, Value addOffset) {
-      curOperand.set(nonUniformOffset);
-  };
+// Offset extraction logic for an addition op:
+// decompose(A+B) = {U(A)+U(B), NU(A)+NU(B)}
+std::pair<Value, Value>
+PointerCanonicalizer::decomposeOffsetFromAdd(Location loc, Value expr,
+                                             int64_t bitness) {
+  auto addOp = cast<arith::AddIOp>(expr.getDefiningOp());
+  auto [uniformOffsetL, nonUniformOffsetL] =
+      decomposeOffsetFromExpr(loc, addOp.getLhs(), bitness);
+  auto [uniformOffsetR, nonUniformOffsetR] =
+      decomposeOffsetFromExpr(loc, addOp.getRhs(), bitness);
+  Value uniformAdd =
+      rewriter.create<arith::AddIOp>(loc, uniformOffsetL, uniformOffsetR);
+  Value nonUniformAdd =
+      rewriter.create<arith::AddIOp>(loc, nonUniformOffsetL, nonUniformOffsetR);
+  return {uniformAdd, nonUniformAdd};
+}
 
-  // If one out lhs or rhs is a scalar, set the current operand to the
-  // non-scalar value and keep extracting
-  if (Value lhsScalar = getScalarConstant(rewriter, loc, lhsOp)) {
-    propagateNonUniformOffset(rhsOperand.get());
-    Value scalarOffset = extractScalarOffsetFromExpr(
-        loc, rhsOp, rhsOperand, bitness, hasNonUniformComponent);
-    return createBinOp(scalarOffset, lhsScalar);
-  } else if (Value rhsScalar = getScalarConstant(rewriter, loc, rhsOp)) {
-    propagateNonUniformOffset(lhsOperand.get());
-    Value scalarOffset = extractScalarOffsetFromExpr(
-        loc, lhsOp, lhsOperand, bitness, hasNonUniformComponent);
-    return createBinOp(scalarOffset, rhsScalar);
-  } else if (isa<arith::MulIOp>(binOp)) {
-    // Stop if this is a non-scalar * non-scalar
-    hasNonUniformComponent = true;
-    return rewriter.create<arith::ConstantIntOp>(loc, 0, bitness);
-  } else {
-    // Keep extracting otherwise
-    Value scalarOffsetLhs = extractScalarOffsetFromExpr(
-        loc, lhsOp, lhsOperand, bitness, hasNonUniformComponent);
-    Value scalarOffsetRhs = extractScalarOffsetFromExpr(
-        loc, rhsOp, rhsOperand, bitness, hasNonUniformComponent);
-    return createBinOp(scalarOffsetLhs, scalarOffsetRhs);
-  }
+// Offset extraction logic for a multiplication op:
+// decompose(A*B) = {U(A)*U(B), NU(A)*NU(B)+NU(B)*U(A)+U(A)*NU(B)}
+std::pair<Value, Value>
+PointerCanonicalizer::decomposeOffsetFromMul(Location loc, Value expr,
+                                             int64_t bitness) {
+  auto mulOp = cast<arith::MulIOp>(expr.getDefiningOp());
+  auto [uniformOffsetL, nonUniformOffsetL] =
+      decomposeOffsetFromExpr(loc, mulOp.getLhs(), bitness);
+  auto [uniformOffsetR, nonUniformOffsetR] =
+      decomposeOffsetFromExpr(loc, mulOp.getRhs(), bitness);
+  Value uniformMul =
+      rewriter.create<arith::MulIOp>(loc, uniformOffsetL, uniformOffsetR);
+
+  Value uniformOffsetLSplat = rewriter.create<triton::SplatOp>(
+      loc, nonUniformOffsetL.getType(), uniformOffsetL);
+  Value uniformOffsetRSplat = rewriter.create<triton::SplatOp>(
+      loc, nonUniformOffsetL.getType(), uniformOffsetR);
```

Comment:
Nit: it might not matter but still better to use `nonUniformOffsetR.getType()` here for consistency